SurfaceDevCenterManager/Utility/AuthorizationHandler.cs

```diff
@@ -31,53 +33,63 @@ public AuthorizationHandler(AuthorizationHandlerCredentials credentials)
             if (_AccessToken == null)
             {
                 await ObtainAccessToken();
             }
-            request.Headers.Add("Authorization", "Bearer " + _AccessToken);
-
-            //Send request
-            HttpResponseMessage response = await base.SendAsync(request, cancellationToken);
 
-            //If unauthorized, the token likely expired so get a new one
-            if (response.StatusCode == HttpStatusCode.Unauthorized)
+            while (tries < MAX_RETRIES)
             {
+                tries++;
+
+                //Clone the original request so we have a copy in case of a failure
+                HttpRequestMessage clonedRequest = await CloneHttpRequestMessageAsync(request);
+
+                clonedRequest.Headers.Add("Authorization", "Bearer " + _AccessToken);
+
+                //Send request
                 try
                 {
-                    //Get a new access token
-                    if (await ObtainAccessToken())
-                    {
-                        //Add token to cloned request and resend
-                        clonedRequest.Headers.Add("Authorization", "Bearer " + _AccessToken);
-                        response = await base.SendAsync(clonedRequest, cancellationToken);
-                    }
+                    response = await base.SendAsync(clonedRequest, cancellationToken);
                 }
-                catch (InvalidOperationException)
+                catch (System.Net.Sockets.SocketException)
                 {
-                    return response;
+                    //HDC timed out, wait a bit and try again
+                    Thread.Sleep(2000);
+                    continue;
                 }
+
+                //If unauthorized, the token likely expired so get a new one and retry
+                if (response.StatusCode == HttpStatusCode.Unauthorized)
+                {
+                    await ObtainAccessToken();
+                    continue;
+                }
+                else if (response.StatusCode == HttpStatusCode.InternalServerError)
+                {
+                    //Somtimes HDC returns 500 errors so wait a bit then retry once instead of failing the whole flow
+                    Thread.Sleep(2000);
+                    continue;
+                }
+
+                break;
             }
-            else if (response.StatusCode == HttpStatusCode.InternalServerError)
-            {
-                //Somtimes HWDC returns 500 errors so wait a bit then retry once instead of failing the whole flow
-                Thread.Sleep(2000);
 
-                // Resend the request with the token
-                clonedRequest.Headers.Add("Authorization", "Bearer " + _AccessToken);
-                response = await base.SendAsync(clonedRequest, cancellationToken);
+            if (response == null)
+            {
+                throw new ApplicationException("AuthorizationHandler: NULL response, unable to talk to HDC");
```

Comment:
Just read recently that ApplicationException is not recommended to use:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/using-standard-exception-types

Reply:
I will find a better Exception to use in a future update - thank you!